core/src/edu/cornell/gdiac/main/GDXRoot.java

```diff
@@ -115,21 +134,122 @@ public void resize(int width, int height) {
    * @param screen   The screen requesting to exit
    * @param exitCode The state of the screen upon exit
    */
+  @Override
   public void exitScreen(Screen screen, int exitCode) {
     if (screen == loading) {
-      directory = loading.getAssets();
-      controller.gatherAssets(directory);
-      controller.setScreenListener(this);
-      controller.setCanvas(canvas);
-      controller.reset();
-      setScreen(controller);
-
-      loading.dispose();
-      loading = null;
-    } else if (exitCode == GameController.EXIT_QUIT) {
-      // We quit the main application
+      exitLoadingScreen(exitCode);
+    } else if (screen == levelSelector) {
+      exitLevelScreen(exitCode);
+    } else if (screen == controller) {
+      exitGameScreen(exitCode);
+    }
+  }
+
+  /**
+   * Handles exit callback for the loading screen
+   */
+  private void exitLoadingScreen(int exitCode) {
+    directory = loading.getAssets();
+
+    // We can dispose loading since assets are loaded in directory
+    loading.dispose();
+    loading = null;
+
+    // Default navigate to levels screen
+    navigateLevelScreen();
+  }
+
+  /**
+   * Handles exit callback for the level selector.
+   */
+  private void exitLevelScreen(int exitCode) {
+    assert screen == levelSelector;
+    String levelKey = levelSelector.getSelectedLevelKey();
+
+    // Clean up level screen
+    levelSelector.setScreenListener(null);
+
+    if (exitCode == LevelMode.EXIT_TO_GAME) {
+      navigateGameScreen(levelKey);
+    } else if (exitCode == LevelMode.EXIT_QUIT) {
       Gdx.app.exit();
     }
   }
 
+  /**
+   * Handles exit callback for the gameplay controller.
+   */
+  private void exitGameScreen(int exitCode) {
+    assert screen == controller;
+    String levelKey = controller.getGameLevelKey();
+
+    // Clean up game screen for levelKey level
+    controller.setScreenListener(null);
+
+    if (exitCode == GameController.EXIT_TO_LEVELS) {
+      navigateLevelScreen();
+    } else if (exitCode == GameController.EXIT_TO_NEXT_LEVEL) {
+      // Determine next level
+      String nextLevel = findNeighborLevel(levelKey, true);
+
+      navigateGameScreen(nextLevel);
+    } else if (exitCode == GameController.EXIT_TO_PREV_LEVEL) {
+      // Determine previous level
+      String prevLevel = findNeighborLevel(levelKey, false);
+
+      navigateGameScreen(prevLevel);
+    } else if (exitCode == GameController.EXIT_QUIT) {
+      // Default navigate to levels screen
+      navigateLevelScreen();
+//      Gdx.app.exit();
+    }
```

Comment:
I'm a little confused why you navigate to the level screen if the exit code is for quitting.

Reply:
I thought it would be nicer than quitting the entire game and instead quit to the levels screen.